README.md
# group_tableview


flutter 仿iOS TableView 实现分组功能，支持plain和group样式，支持下拉刷新组件

![示例一](https://github.com/GG-beyond/group_tableview/blob/master/images/Untitled1.gif)
![示例二](https://github.com/GG-beyond/group_tableview/blob/master/images/Untitled2.gif)
![示例三](https://github.com/GG-beyond/group_tableview/blob/master/images/Untitled3.gif)
![示例四](https://github.com/GG-beyond/group_tableview/blob/master/images/Untitled4.gif)

CHANGELOG.md
## [0.0.7] -  2020-08-26
* Dart code formatting

## [0.0.6] -  2020-08-24
* Modify the renderings

## [0.0.5] -  2020-08-24
* Fixed group hover bug

## [0.0.4] -  2020-08-24
* Dart development specification

## [0.0.3] -  2020-08-24
* Fixed group hover bug

## [0.0.2] -  2020-08-24
* Fixed group hover bug

## [0.0.1] -  2020-08-24
* Implement ListView group suspension, support refresh component











test/group_listview_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:group_tableview/group_tableview.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
    expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  });
}


example/lib/main.dart
import 'package:example/page1.dart';
import 'package:example/page2.dart';
import 'page3.dart';
import 'page4.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Group TableView',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Group TableView'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  List<String> _list = [
    "table_view plain样式",
    "table_view plain样式 带刷新",
    "table_view group悬停样式",
    "table_view group悬停样式 带刷新"
  ];
  List<Widget> _widgetArr = [Page1(), Page3(), Page2(), Page4()];

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Container(
        color: Colors.white,
        child: ListView.builder(
          itemBuilder: (context, index) {
            return _itemBuilder(context, index);
          },
          itemCount: _list.length,
        ),
      ),
    );
  }

  Widget _itemBuilder(BuildContext context, int index) {
    return InkWell(
      child: Padding(
        padding: const EdgeInsets.symmetric(horizontal: 8.0),
        child: Card(
          elevation: 8,
          child: ListTile(
            contentPadding:
                const EdgeInsets.symmetric(horizontal: 18, vertical: 10.0),
            title: Text(
              _list[index],
              style: TextStyle(fontSize: 16, fontWeight: FontWeight.w400),
            ),
            trailing: Icon(Icons.arrow_forward_ios),
          ),
        ),
      ),
      onTap: () {
        Widget page = _widgetArr[index];
        Navigator.of(context)
            .push(CupertinoPageRoute(builder: (BuildContext context) {
          return page;
        }));
      },
    );
  }
}


example/lib/page1.dart
import 'package:flutter/material.dart';
import 'package:group_tableview/group_tableview.dart';
import 'package:pull_to_refresh/pull_to_refresh.dart';
import 'package:flutter/cupertino.dart';

class Page1 extends StatefulWidget {
  @override
  _Page1State createState() => _Page1State();
}

class _Page1State extends State<Page1> {
  List<Map> _list = [
    {
      "group": "A",
      "list": ["Apple", "安琪拉", "A-桔子🍊🍊🍊", "艾菲尔"]
    },
    {
      "group": "B",
      "list": ["拜仁", "白起", "Boy", "宝丁"]
    },
    {
      "group": "C",
      "list": ["城主", "Channel", "长安城管", "Charles", "长江"]
    },
    {
      "group": "D",
      "list": ["东南西北"]
    },
  ];

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text("TableView Plain"),
      ),
      body: Container(
        child: GroupTableView(
          style: ViewStyle.plain,
          itemBuilder: _itemBuilder,
          numberOfSections: _list.length,
          numberOfRowsInSection: (int section) {
            return _list[section]["list"].length;
          },
          sectionFooterBuilder: (context, section) {
            Color bgColor = section % 2 == 0 ? Colors.red : Colors.green;
            return Container(
              color: bgColor,
              child: Padding(
                padding: EdgeInsets.only(left: 20, top: 10, bottom: 10),
                child: Text(
                    "              这是一个footer -> section = ${section.toString()}"),
              ),
            );
          },
          sectionHeaderBuilder: (BuildContext context, int section) {
            Color bgColor = section % 2 == 0 ? Colors.red : Colors.green;
            String title = _list[section]["group"];
            double height = section % 2 == 0 ? 40 : 60;

            return Container(
              height: height,
              color: bgColor,
              child: Padding(
                padding:
                    const EdgeInsets.symmetric(horizontal: 15, vertical: 8),
                child: Text(
                  "$title             这是一个header -> section = ${section.toString()}",
                  style: TextStyle(fontSize: 18, fontWeight: FontWeight.w600),
                ),
              ),
            );
          },
        ),
      ),
    );
  }

  Widget _itemBuilder(BuildContext context, IndexPath indexPath) {
    return Padding(
      padding: const EdgeInsets.symmetric(horizontal: 8.0),
      child: Card(
        elevation: 8,
        child: ListTile(
          contentPadding:
              const EdgeInsets.symmetric(horizontal: 18, vertical: 10.0),
          title: Text(
            _list[indexPath.section]["list"][indexPath.row],
            style: TextStyle(fontSize: 16, fontWeight: FontWeight.w400),
          ),
          subtitle: Text(
              "section = ${indexPath.section.toString()},row = ${indexPath.row.toString()}"),
          trailing: Icon(Icons.arrow_forward_ios),
        ),
      ),
    );
  }
}


example/lib/page2.dart
import 'package:flutter/material.dart';
import 'package:group_tableview/group_tableview.dart';
import 'package:pull_to_refresh/pull_to_refresh.dart';
import 'package:flutter/cupertino.dart';

class Page2 extends StatefulWidget {
  @override
  _Page2State createState() => _Page2State();
}

class _Page2State extends State<Page2> {
  ScrollController _controller = ScrollController();

  List<Map> _list = [
    {
      "group": "A",
      "list": ["Apple", "安琪拉", "A-桔子🍊🍊🍊", "艾菲尔"]
    },
    {
      "group": "B",
      "list": ["拜仁", "白起", "Boy", "宝丁"]
    },
    {
      "group": "C",
      "list": ["城主", "Channel", "长安城管", "Charles", "长江"]
    },
    {
      "group": "D",
      "list": ["东南西北"]
    },
  ];

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text("TableView Group"),
      ),
      body: Container(
        child: GroupTableView(
          controller: _controller,
          style: ViewStyle.group,
          itemBuilder: _itemBuilder,
          numberOfSections: _list.length,
          numberOfRowsInSection: (int section) {
            return _list[section]["list"].length;
          },
          sectionFooterBuilder: (context, section) {
            Color bgColor = section % 2 == 0 ? Colors.red : Colors.green;
            return Container(
              color: bgColor,
              child: Padding(
                padding: EdgeInsets.only(left: 20, top: 10, bottom: 10),
                child: Text(
                    "              这是一个footer -> section = ${section.toString()}"),
              ),
            );
          },
          sectionHeaderBuilder: (BuildContext context, int section) {
            Color bgColor = section % 2 == 0 ? Colors.red : Colors.green;
            String title = _list[section]["group"];
            double height = section % 2 == 0 ? 40 : 60;
            return Container(
              height: height,
              color: bgColor,
              child: Padding(
                padding:
                    const EdgeInsets.symmetric(horizontal: 15, vertical: 8),
                child: Text(
                  "$title             这是一个header -> section = ${section.toString()}",
                  style: TextStyle(fontSize: 18, fontWeight: FontWeight.w600),
                ),
              ),
            );
          },
        ),
      ),
    );
  }

  Widget _itemBuilder(BuildContext context, IndexPath indexPath) {
    return Padding(
      padding: const EdgeInsets.symmetric(horizontal: 8.0),
      child: Card(
        elevation: 8,
        child: ListTile(
          contentPadding:
              const EdgeInsets.symmetric(horizontal: 18, vertical: 10.0),
          title: Text(
            _list[indexPath.section]["list"][indexPath.row],
            style: TextStyle(fontSize: 16, fontWeight: FontWeight.w400),
          ),
          subtitle: Text(
              "section = ${indexPath.section.toString()},row = ${indexPath.row.toString()}"),
          trailing: Icon(Icons.arrow_forward_ios),
        ),
      ),
    );
  }
}


example/lib/page3.dart
import 'package:flutter/material.dart';
import 'package:group_tableview/group_tableview.dart';
import 'package:pull_to_refresh/pull_to_refresh.dart';
import 'package:flutter/cupertino.dart';

class Page3 extends StatefulWidget {
  @override
  _Page3State createState() => _Page3State();
}

class _Page3State extends State<Page3> {
  RefreshController _refreshController;
  List<Map> _list = [
    {
      "group": "A",
      "list": ["Apple", "安琪拉", "A-桔子🍊🍊🍊", "艾菲尔"]
    },
    {
      "group": "B",
      "list": ["拜仁", "白起", "Boy", "宝丁"]
    },
    {
      "group": "C",
      "list": ["城主", "Channel", "长安城管", "Charles", "长江"]
    },
    {
      "group": "D",
      "list": ["东南西北"]
    },
  ];

  void _onRefresh() async {
    await Future.delayed(Duration(milliseconds: 1000));
    _refreshController.refreshCompleted();
  }

  void _onLoading() async {
    await Future.delayed(Duration(milliseconds: 1000));
    _refreshController.loadComplete();
  }

  @override
  void initState() {
    // TODO: implement initState
    super.initState();
    _refreshController = RefreshController();
  }

  @override
  void dispose() {
    // TODO: implement dispose
    super.dispose();
    _refreshController.dispose();
  }

  @override
  Widget build(BuildContext context) {
    Widget _refreshWidget(BuildContext context, Widget listView) {
      return SmartRefresher(
        enablePullDown: true,
        enablePullUp: true,
        header: WaterDropHeader(),
        footer: CustomFooter(
          builder: (BuildContext context, LoadStatus mode) {
            Widget body;
            if (mode == LoadStatus.idle) {
              body = Text("上拉加载");
            } else if (mode == LoadStatus.loading) {
              body = CupertinoActivityIndicator();
            } else if (mode == LoadStatus.failed) {
              body = Text("加载失败！点击重试！");
            } else if (mode == LoadStatus.canLoading) {
              body = Text("松手,加载更多!");
            } else {
              body = Text("没有更多数据了!");
            }
            return Container(
              height: 55.0,
              child: Center(child: body),
            );
          },
        ),
        controller: _refreshController,
        onRefresh: _onRefresh,
        onLoading: _onLoading,
        child: listView,
      );
    }

    return Scaffold(
      appBar: AppBar(
        title: Text("TableView Plain"),
      ),
      body: Container(
        child: GroupTableView(
          style: ViewStyle.plain,
          itemBuilder: _itemBuilder,
          numberOfSections: _list.length,
          numberOfRowsInSection: (int section) {
            return _list[section]["list"].length;
          },
          sectionFooterBuilder: (context, section) {
            Color bgColor = section % 2 == 0 ? Colors.red : Colors.green;
            return Container(
              color: bgColor,
              child: Padding(
                padding: EdgeInsets.only(left: 20, top: 10, bottom: 10),
                child: Text(
                    "              这是一个footer -> section = ${section.toString()}"),
              ),
            );
          },
          sectionHeaderBuilder: (BuildContext context, int section) {
            Color bgColor = section % 2 == 0 ? Colors.red : Colors.green;
            String title = _list[section]["group"];
            double height = section % 2 == 0 ? 40 : 60;

            return Container(
              height: height,
              color: bgColor,
              child: Padding(
                padding:
                    const EdgeInsets.symmetric(horizontal: 15, vertical: 8),
                child: Text(
                  "$title             这是一个header -> section = ${section.toString()}",
                  style: TextStyle(fontSize: 18, fontWeight: FontWeight.w600),
                ),
              ),
            );
          },
          refreshWidget: _refreshWidget,
        ),
      ),
    );
  }

  Widget _itemBuilder(BuildContext context, IndexPath indexPath) {
    return Padding(
      padding: const EdgeInsets.symmetric(horizontal: 8.0),
      child: Card(
        elevation: 8,
        child: ListTile(
          contentPadding:
              const EdgeInsets.symmetric(horizontal: 18, vertical: 10.0),
          title: Text(
            _list[indexPath.section]["list"][indexPath.row],
            style: TextStyle(fontSize: 16, fontWeight: FontWeight.w400),
          ),
          subtitle: Text(
              "section = ${indexPath.section.toString()},row = ${indexPath.row.toString()}"),
          trailing: Icon(Icons.arrow_forward_ios),
        ),
      ),
    );
  }
}


lib/src/mylistview.dart
import 'package:flutter/material.dart';

///
/// 类型 IndexPath
///
class IndexPath {
  final int section;
  final int row;

  IndexPath(this.section, this.row);
}

///
/// 类型 模仿iOS tableview样式，默认是plain
///
enum ViewStyle {
  plain, //regular table view
  group //sections are grouped together
}

///
/// 定义方法
///
typedef IndexPathWidgetBuilder = Widget Function(
    BuildContext context, IndexPath indexPath);
typedef SectionBuilder = int Function(int section);
typedef Widget ListViewSuperWidgetBuilder(
    BuildContext context, Widget scrollWidget);

///
/// GroupTableView 实现clean
///
class GroupTableView extends StatefulWidget {
  final ViewStyle style; //listview样式,默认是plain样式
  final int numberOfSections; //section个数 默认1
  final SectionBuilder numberOfRowsInSection; //每个section内有多少个row
  final IndexPathWidgetBuilder itemBuilder; //item builder方法
  final IndexedWidgetBuilder sectionHeaderBuilder; // section header builder 方法
  final IndexedWidgetBuilder sectionFooterBuilder; // section footer builder 方法
  final ListViewSuperWidgetBuilder refreshWidget; //父 刷新组件
  final ScrollController controller; // controller
  final Color backgroundColor; // 背景颜色

  GroupTableView(
      {this.itemBuilder,
      this.style = ViewStyle.plain,
      this.numberOfSections = 1,
      this.numberOfRowsInSection,
      this.sectionHeaderBuilder,
      this.sectionFooterBuilder,
      this.refreshWidget,
      this.controller,
      this.backgroundColor})
      : assert(itemBuilder != null, "itemBuilder 不能为null");

  @override
  _GroupTableViewState createState() => _GroupTableViewState();
}

class _GroupTableViewState extends State<GroupTableView> {
  ScrollController _controller;
  ListView listView;
  List<SectionModel> sectionList = [];
  List<GlobalKey> keyList = [];
  SectionModel currentSectionModel;
  double position = -88; //悬浮位置(相对于设备)
  int currentIndex = -1; //当前是第几个分区悬浮
  double topOffsetY = 0;
  bool floating = true;

  Size _size;

  static int totalCount = 0;

  @override
  void initState() {
    // TODO: implement initState
    super.initState();

    _controller = widget.controller;
    if (_controller == null) {
      _controller = ScrollController();
    }
    if (widget.style == ViewStyle.group) {
      _controller.addListener(() {
        double offsetY = _controller.offset;
        int _currentIndex = -1;

        if (offsetY <= 0.0) {
          currentIndex = -1; //当前是第几个分区悬浮
          floating = false;
          setState(() {});
        } else {
          floating = true;
          for (int section = sectionList.length - 1; section >= 0; section--) {
            GlobalKey globalKey = keyList[section];
            if (globalKey.currentContext != null) {
              RenderBox renderBox = globalKey.currentContext.findRenderObject();
              double dy = renderBox.localToGlobal(Offset(0, position)).dy;

              if (dy <= 0) {
                //查找最后一个悬浮的
                _currentIndex = section;
                _size = renderBox.size;
                sectionList[section].headerHeight = _size.height;
                break;
              }
            }
          }

          //加一层复用保护
          if (_currentIndex < 0 && currentIndex >= 0) {
            GlobalKey key = keyList[currentIndex];
            if (key.currentContext == null) {
              _currentIndex = currentIndex;
              _size = Size(_size.width, sectionList[currentIndex].headerHeight);
            } else {
              _currentIndex = currentIndex - 1;
            }
          }

          double _offset = 0;

          if (_currentIndex >= 0) {
            if ((_currentIndex + 1) < sectionList.length) {
              //当前悬浮的
              GlobalKey nextGlobalKey = keyList[_currentIndex + 1];

              if (nextGlobalKey.currentContext != null) {
                RenderBox nextRenderBox =
                    nextGlobalKey.currentContext.findRenderObject();
                double nextDy =
                    nextRenderBox.localToGlobal(Offset(0, position)).dy;

                //取出滚动方向对应的数据
                double offsetAxis = nextDy;
                double sizeAxis = _size.height;

                //计算偏移位置
                if (offsetAxis < sizeAxis) {
                  _offset = offsetAxis - sizeAxis;
                }
              }
            }
          }

          if (_currentIndex != currentIndex || _offset != topOffsetY) {
            if (_currentIndex > -1) {
              currentIndex = _currentIndex;
              currentSectionModel = sectionList[_currentIndex];
              topOffsetY = _offset;
            }
            setState(() {});
          }
        }
      });
    }

    _initData();
  }

  ///根据DataSource 初始化数据
  void _initData() {
    totalCount = 0;
    sectionList.clear();

    for (int section = 0; section < widget.numberOfSections; section++) {
      //遍历section
      int rowCount =
          widget.numberOfRowsInSection(section); //获取每个section(区)有多少个row

      Widget header; //获取header
      if (widget.sectionHeaderBuilder != null) {
        header = widget.sectionHeaderBuilder(context, section);
      }
      Widget footer; //获取footer
      if (widget.sectionFooterBuilder != null) {
        footer = widget.sectionFooterBuilder(context, section);
      }

      bool isHaveHeader = header != null ? true : false; //是否有header
      bool isHaveFooter = footer != null ? true : false; //是否有footer

      SectionModel sectionModel =
          SectionModel(section, rowCount, isHaveHeader, isHaveFooter, header);
      if (section == 0) {
        currentSectionModel = sectionModel;
      }
      sectionList.add(sectionModel);

      totalCount = totalCount +
          rowCount +
          (header != null ? 1 : 0) +
          (footer != null ? 1 : 0); //ListView itemCount总行数

      GlobalKey globalKey = GlobalKey(debugLabel: section.toString());
      keyList.add(globalKey);
    }
    listView = ListView.builder(
      controller: _controller,
      physics: BouncingScrollPhysics(),
      itemBuilder: _itemBuilder,
      itemCount: _calculateItemCount(),
    );
  }

  @override
  Widget build(BuildContext context) {
    Widget child;
    if (widget.refreshWidget != null &&
        widget.refreshWidget(context, listView) != null) {
      child = widget.refreshWidget(context, listView);
    } else {
      child = listView;
    }

    if (widget.style == ViewStyle.plain) {
      return Container(
        color: widget.backgroundColor ?? Colors.transparent,
        child: child,
      );
    }
    return Container(
      child: Stack(
        children: <Widget>[
          Container(
            color: widget.backgroundColor ?? Colors.transparent,
            child: child,
          ),
          Visibility(
            visible: floating,
            child: Positioned(
                left: 0,
                right: 0,
                top: topOffsetY ?? 0,
                child: currentSectionModel.header ?? Container()),
          ),
        ],
      ),
    );
  }

  int _calculateItemCount() {
    return totalCount; //row+section = 总行数
  }

  Widget _itemBuilder(BuildContext context, int index) {
    Widget item;
    Object model = _getItemRowModel(index);

    if (model is SectionHeaderModel) {
      SectionHeaderModel sectionHeaderModel = model;
      item = Container(
        key: keyList[sectionHeaderModel.section],
        child: widget.sectionHeaderBuilder(context, sectionHeaderModel.section),
      );
//      item = widget.sectionHeaderBuilder(context, sectionHeaderModel.section);
    } else if (model is SectionFooterModel) {
      SectionFooterModel sectionFooterModel = model;
      item = widget.sectionFooterBuilder(context, sectionFooterModel.section);
    } else {
      RowModel rowModel = model;
      item = widget.itemBuilder(context, rowModel.indexPath);
    }
    return item;
  }

  Object _getItemRowModel(int index) {
    int passCount = 0;
    //遍历整个分区 ，去查找index在哪个分区section的哪个下标row
    for (int section = 0; section < sectionList.length; section++) {
      SectionModel sectionModel = this.sectionList[section];

      bool isHaveHeader = sectionModel.isHaveHeader;
      bool isHaveFooter = sectionModel.isHaveFooter;

      int tempCount = 0;
      if (isHaveHeader == true) {
        //有header
        tempCount = tempCount + 1;
      }

      if (index == passCount && isHaveHeader == true) {
        //这是header
        return SectionHeaderModel(section);
      } else if (index == tempCount + sectionModel.rowCount + passCount &&
          isHaveFooter == true) {
        //这是footer
        return SectionFooterModel(section);
      } else if (index >= passCount &&
          index < tempCount + sectionModel.rowCount + passCount) {
        //这是row
        IndexPath indexPath = IndexPath(section, index - passCount - tempCount);
        return RowModel(indexPath);
      }
      passCount = passCount +
          sectionModel.rowCount +
          tempCount +
          (isHaveFooter ? 1 : 0);
    }
    return null;
  }
}

///
///  ============================ model类 ===============================================
///
class SectionModel {
  final Widget header; //是否有header
  final int section; //当前section
  final int rowCount; //记录当前section下有多少个row
  final bool isHaveHeader; //是否有header
  final bool isHaveFooter; //是否有footer

  SectionModel(this.section, this.rowCount, this.isHaveHeader,
      this.isHaveFooter, this.header);

  double headerHeight;
}

///
/// 记录section header 的类 包含区域
///
class SectionHeaderModel {
  final int section; //当前section
  SectionHeaderModel(this.section);
}

///
/// 记录section footer 的类 包含区域
///
class SectionFooterModel {
  final int section; //当前section
  SectionFooterModel(this.section);
}

///
///记录 row 的model类，
///
class RowModel {
  final IndexPath indexPath; //indexPath 的section和row
  RowModel(this.indexPath);
}


example/lib/page4.dart
import 'package:flutter/material.dart';
import 'package:group_tableview/group_tableview.dart';
import 'package:pull_to_refresh/pull_to_refresh.dart';
import 'package:flutter/cupertino.dart';

class Page4 extends StatefulWidget {
  @override
  _Page4State createState() => _Page4State();
}

class _Page4State extends State<Page4> {
  RefreshController _refreshController;
  ScrollController _controller = ScrollController();

  List<Map> _list = [
    {
      "group": "A",
      "list": ["Apple", "安琪拉", "A-桔子🍊🍊🍊", "艾菲尔"]
    },
    {
      "group": "B",
      "list": ["拜仁", "白起", "Boy", "宝丁"]
    },
    {
      "group": "C",
      "list": ["城主", "Channel", "长安城管", "Charles", "长江"]
    },
    {
      "group": "D",
      "list": ["东南西北"]
    },
  ];

  void _onRefresh() async {
    await Future.delayed(Duration(milliseconds: 1000));
    _refreshController.refreshCompleted();
  }

  void _onLoading() async {
    await Future.delayed(Duration(milliseconds: 1000));
    _refreshController.loadComplete();
  }

  @override
  void initState() {
    // TODO: implement initState
    super.initState();
    _refreshController = RefreshController();
  }

  @override
  void dispose() {
    // TODO: implement dispose
    super.dispose();
    _refreshController.dispose();
    _controller.dispose();
  }

  @override
  Widget build(BuildContext context) {
    Widget _refreshWidget(BuildContext context, Widget listView) {
      return SmartRefresher(
        scrollController: _controller,
        physics: BouncingScrollPhysics(),
        enablePullDown: true,
        enablePullUp: true,
        header: WaterDropHeader(),
        footer: CustomFooter(
          builder: (BuildContext context, LoadStatus mode) {
            Widget body;
            if (mode == LoadStatus.idle) {
              body = Text("上拉加载");
            } else if (mode == LoadStatus.loading) {
              body = CupertinoActivityIndicator();
            } else if (mode == LoadStatus.failed) {
              body = Text("加载失败！点击重试！");
            } else if (mode == LoadStatus.canLoading) {
              body = Text("松手,加载更多!");
            } else {
              body = Text("没有更多数据了!");
            }
            return Container(
              height: 55.0,
              child: Center(child: body),
            );
          },
        ),
        controller: _refreshController,
        onRefresh: _onRefresh,
        onLoading: _onLoading,
        child: listView,
      );
    }

    return Scaffold(
      appBar: AppBar(
        title: Text("TableView Group"),
      ),
      body: Container(
        child: GroupTableView(
          controller: _controller,
          style: ViewStyle.group,
          itemBuilder: _itemBuilder,
          numberOfSections: _list.length,
          numberOfRowsInSection: (int section) {
            return _list[section]["list"].length;
          },
          sectionFooterBuilder: (context, section) {
            Color bgColor = section % 2 == 0 ? Colors.red : Colors.green;
            return Container(
              color: bgColor,
              child: Padding(
                padding: EdgeInsets.only(left: 20, top: 10, bottom: 10),
                child: Text(
                    "              这是一个footer -> section = ${section.toString()}"),
              ),
            );
          },
          sectionHeaderBuilder: (BuildContext context, int section) {
            Color bgColor = section % 2 == 0 ? Colors.red : Colors.green;
            String title = _list[section]["group"];
            double height = section % 2 == 0 ? 40 : 60;
            return Container(
              height: height,
              color: bgColor,
              child: Padding(
                padding:
                    const EdgeInsets.symmetric(horizontal: 15, vertical: 8),
                child: Text(
                  "$title             这是一个header -> section = ${section.toString()}",
                  style: TextStyle(fontSize: 18, fontWeight: FontWeight.w600),
                ),
              ),
            );
          },
          refreshWidget: _refreshWidget,
        ),
      ),
    );
  }

  Widget _itemBuilder(BuildContext context, IndexPath indexPath) {
    return Padding(
      padding: const EdgeInsets.symmetric(horizontal: 8.0),
      child: Card(
        elevation: 8,
        child: ListTile(
          contentPadding:
              const EdgeInsets.symmetric(horizontal: 18, vertical: 10.0),
          title: Text(
            _list[indexPath.section]["list"][indexPath.row],
            style: TextStyle(fontSize: 16, fontWeight: FontWeight.w400),
          ),
          subtitle: Text(
              "section = ${indexPath.section.toString()},row = ${indexPath.row.toString()}"),
          trailing: Icon(Icons.arrow_forward_ios),
        ),
      ),
    );
  }
}
